src/components/PupilComponents/LessonEngineProvider/LessonEngineProvider.tsx

```diff
@@ -150,16 +150,22 @@ export const useLessonEngineContext = () => {
 export type LessonEngineProviderProps = {
   children: ReactNode;
   initialLessonReviewSections: Readonly<LessonReviewSection[]>;
+  sectionCompletedTrackingCB?: (section: LessonReviewSection) => void;
 };
 
 export const LessonEngineProvider = memo(
-  ({ children, initialLessonReviewSections }: LessonEngineProviderProps) => {
+  ({
+    children,
+    initialLessonReviewSections,
+    sectionCompletedTrackingCB,
+  }: LessonEngineProviderProps) => {
     const [state, dispatch] = useReducer(lessonEngineReducer, {
       lessonReviewSections: initialLessonReviewSections,
       currentSection: "overview",
       sections: {},
     });
     const completeSection = (section: LessonReviewSection) => {
+      if (sectionCompletedTrackingCB) sectionCompletedTrackingCB(section);
       dispatch({ type: "completeSection", section });
     };
```

Comment:
tbh I wouldn't abstract at this point and drilling the callback down through the provider is a lot of boilerplate — especially if you then need to do it for every action. I think it's ok to call the tracking directly since `LessonEngineProvider` doesn't have any use elsewhere. If it were more generic we might consider abstracting the tracking away — perhaps with some middleware that intercepts `dispatch` or something.

I'm guessing you'll be using the existing analytics provider and hook

```typescript
const { track } = useAnalytics(); 

...

const completeSection = (section: LessonReviewSection) => {
  dispatch({ type: "completeSection", section });
  track.pupilLessonSectionComplete(section);
}
```

Easy to follow what's going on and it's already separate from the UI.

Also i'd always put the tracking after the dispatch in case it blows up out of an abundance of caution and mistrust of analytics products.